File: extension/core/src/main/java/io/holunda/camunda/bpm/data/reader/TaskServiceVariableReader.java
package io.holunda.camunda.bpm.data.reader;

import io.holunda.camunda.bpm.data.factory.VariableFactory;
import org.camunda.bpm.engine.TaskService;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;
import java.util.Optional;

/**
 * Allows reading multiple variable values from {@link TaskService#getVariable(String, String)}.
 */
public class TaskServiceVariableReader implements VariableReader {

  private final TaskService taskService;
  private final String taskId;

  /**
   * Constructs a reader operating on task service.
   *
   * @param taskService task service to operate on.
   * @param taskId      task id.
   */
  public TaskServiceVariableReader(TaskService taskService, String taskId) {
    this.taskService = taskService;
    this.taskId = taskId;
  }


  @NotNull
  @Override
  public <T> Optional<T> getOptional(VariableFactory<T> variableFactory) {
    return variableFactory.from(taskService, taskId).getOptional();
  }

  @NotNull
  @Override
  public <T> T get(VariableFactory<T> variableFactory) {
    return variableFactory.from(taskService, taskId).get();
  }

  @NotNull
  @Override
  public <T> T getLocal(VariableFactory<T> variableFactory) {
    return variableFactory.from(taskService, taskId).getLocal();
  }

  @NotNull
  @Override
  public <T> Optional<T> getLocalOptional(VariableFactory<T> variableFactory) {
    return variableFactory.from(taskService, taskId).getLocalOptional();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    TaskServiceVariableReader that = (TaskServiceVariableReader) o;

    if (!Objects.equals(taskService, that.taskService)) return false;
    return Objects.equals(taskId, that.taskId);
  }

  @Override
  public int hashCode() {
    int result = taskService != null ? taskService.hashCode() : 0;
    result = 31 * result + (taskId != null ? taskId.hashCode() : 0);
    return result;
  }
}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/reader/TaskServiceVariableReader.kt
package io.holunda.camunda.bpm.data.reader

import io.holunda.camunda.bpm.data.factory.VariableFactory
import org.camunda.bpm.engine.TaskService
import java.util.*

/**
 * Allows reading multiple variable values from [TaskService.getVariable].
 * @param taskService task service to operate on.
 * @param taskId      task id.
 */
class TaskServiceVariableReader(private val taskService: TaskService, private val taskId: String) : VariableReader {
    override fun <T> getOptional(variableFactory: VariableFactory<T>): Optional<T> {
        return variableFactory.from(taskService, taskId).getOptional()
    }

    override fun <T> get(variableFactory: VariableFactory<T>): T {
        return variableFactory.from(taskService, taskId).get()
    }

    override fun <T> getLocal(variableFactory: VariableFactory<T>): T {
        return variableFactory.from(taskService, taskId).getLocal()
    }

    override fun <T> getLocalOptional(variableFactory: VariableFactory<T>): Optional<T> {
        return variableFactory.from(taskService, taskId).getLocalOptional()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as TaskServiceVariableReader
        return if (taskService != that.taskService) false else taskId == that.taskId
    }

    override fun hashCode(): Int {
      return Objects.hash(taskService, taskId)
    }
}
